pkg/common/cns-lib/volume/manager.go

```diff
@@ -54,19 +54,34 @@ const (
 	taskInvocationStatusInProgress = cnsvolumeoperationrequest.TaskInvocationStatusInProgress
 	taskInvocationStatusSuccess    = cnsvolumeoperationrequest.TaskInvocationStatusSuccess
 	taskInvocationStatusError      = cnsvolumeoperationrequest.TaskInvocationStatusError
+
+	// CSITaskInfoEmptyFault is the fault type when taskInfo is empty.
+	CSITaskInfoEmptyFault = "csi.fault.TaskInfoEmpty"
+
+	// CSITaskResultEmptyFault is the fault type when taskResult is empty.
+	CSITaskResultEmptyFault = "csi.fault.TaskResultEmpty"
+
+	// CSIInternalFault is the fault type returned when CSI internal error occurs.
+	CSIInternalFault = "csi.fault.Internal"
+	// CSINotFoundFault is the fault type returned when object required is not found.
+	CSINotFoundFault = "csi.fault.NotFound"
+	// CNSInvalidArgumentFault is the fault type returned when invalid argument is given.
+	CSIInvalidArgumentFault = "csi.fault.InvalidArgument"
+	// CSIUnimplementedFault is the fault type returned when the function is unimplemented.
+	CSIUnimplementedFault = "csi.fault.Unimplemented"
 )
 
 // Manager provides functionality to manage volumes.
 type Manager interface {
 	// CreateVolume creates a new volume given its spec.
-	CreateVolume(ctx context.Context, spec *cnstypes.CnsVolumeCreateSpec) (*CnsVolumeInfo, error)
+	CreateVolume(ctx context.Context, spec *cnstypes.CnsVolumeCreateSpec) (*CnsVolumeInfo, string, error)
```

Comment:
This is not intuitive at all, one cannot tell what the string is by just looking at the interface. Furthermore, there is a string and error, both of which provide the same information to the caller, i.e, there was an issue when invoking the API. I think creating a struct with {error, faultType} and returning that instead of a plain error is a better approach.

Reply:
+1